components/Post/PageProgress.tsx

```diff
@@ -3,21 +3,30 @@ import { useEffect, useRef } from "react";
 
 const chromeZoomPixelGapBugFix = -0.25;
 
-const PageProgress = () => {
+interface PageProgressProps {
+  blogPostSectionRef: React.RefObject<HTMLDivElement>;
+}
+
+const PageProgress = ({ blogPostSectionRef }: PageProgressProps) => {
   const progressBarRef = useRef() as React.MutableRefObject<HTMLDivElement>;
 
   useEffect(() => {
+    const blogPostSection = blogPostSectionRef.current;
+
     const calculateScrollProgress = () => {
-      const scrollHeight =
-        document.documentElement.scrollHeight - window.innerHeight;
+      if (!blogPostSection) return;
+
+      const height = blogPostSection.offsetHeight || 0;
+      const offsetTop = blogPostSection.offsetTop || 0;
+      const scrollHeight = height - window.innerHeight;
       const scrollTop =
-        (document.documentElement.scrollTop || document.body.scrollTop) /
-        scrollHeight;
-      const percentage = Math.round(scrollTop * 100);
+        (document.documentElement.scrollTop || document.body.scrollTop) -
+        offsetTop;
+      const percentage = Math.round((scrollTop / scrollHeight) * 100);
 
-      progressBarRef.current.style.setProperty(
+      progressBarRef.current?.style.setProperty(
         "width",
-        percentage > 100 ? "100%" : percentage + "%"
+        percentage > 100 ? "100%" : percentage < 0 ? "0%" : percentage + "%"
```

Comment:
I'm not fun of nested ternaries, can we put the `width` value outside, and calculate it with `if` statements?